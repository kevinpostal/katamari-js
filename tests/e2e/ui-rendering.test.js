/**
 * Snapshot tests for UI rendering and DOM structure
 * Tests DOM structure snapshots for game UI elements, HUD layout consistency, and overlay display structure
 */

import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { setupGameEnvironment, createMockGameState } from '../helpers/game-helpers.js';

describe('UI Rendering Snapshots', () => {
    let gameEnvironment;
    let originalDocument;

    beforeEach(() => {
        // Store original document methods
        originalDocument = {
            createElement: document.createElement.bind(document),
            appendChild: document.body.appendChild.bind(document.body),
            removeChild: document.body.removeChild.bind(document.body)
        };

        // Create the complete game UI structure as it appears in index.html
        document.body.innerHTML = `
            <div id="loading-overlay">Generating a new universe... âœ¨</div>
            <div id="game-ui">
                <div>Size: <span id="katamari-size">2.00m</span></div>
                <div>Speed: <span id="katamari-speed">0.00m/s</span></div>
                <div>Items Collected: <span id="items-collected">0</span></div>
                <div>FPS: <span id="fps">--</span></div>
                <div id="progress-container">
                    <div id="progress-bar"></div>
                </div>
                <div>Target Size: <span id="target-size">100.00m</span></div>
                <div id="power-up-status"></div>
                <button id="gyro-button">Toggle Gyro</button>
                <button id="debug-button">Toggle Debug</button>
            </div>
            <div id="controls-info">
                <p>Use <b>W, A, S, D</b> or <b>Arrow Keys</b> to move the Katamari.</p>
                <p><b>Swipe</b> on screen for precise mobile control.</p>
                <p><b>Toggle Gyro</b> for tilt-based movement on supported devices.</p>
                <p>Press <b>R</b> to reset Katamari position. Press <b>Space</b> to generate a new world.</p>
            </div>
            <div id="message-overlay"></div>
        `;

        // Set up game environment for testing
        gameEnvironment = setupGameEnvironment({
            includePhysics: false,
            includeRendering: false,
            includeAudio: false,
            itemCount: 0
        });

        // Mock window.matchMedia for responsive design tests
        window.matchMedia = vi.fn((query) => ({
            matches: query.includes('pointer: coarse') || query.includes('max-width: 768px'),
            media: query,
            onchange: null,
            addListener: vi.fn(),
            removeListener: vi.fn(),
            addEventListener: vi.fn(),
            removeEventListener: vi.fn(),
            dispatchEvent: vi.fn()
        }));
    });

    afterEach(() => {
        if (gameEnvironment?.dispose) {
            gameEnvironment.dispose();
        }
        
        // Clear DOM
        document.body.innerHTML = '';
        
        // Clear all mocks
        vi.clearAllMocks();
    });

    describe('Game UI Structure Snapshots', () => {
        it('should match complete game UI DOM structure snapshot', () => {
            const gameUI = document.getElementById('game-ui');
            
            // Create a clean representation of the UI structure
            const uiStructure = {
                tagName: gameUI.tagName,
                id: gameUI.id,
                children: Array.from(gameUI.children).map(child => ({
                    tagName: child.tagName,
                    id: child.id || null,
                    className: child.className || null,
                    textContent: child.textContent.trim(),
                    children: child.children.length > 0 ? Array.from(child.children).map(grandchild => ({
                        tagName: grandchild.tagName,
                        id: grandchild.id || null,
                        className: grandchild.className || null,
                        textContent: grandchild.textContent.trim()
                    })) : []
                }))
            };

            expect(uiStructure).toMatchSnapshot('complete-game-ui-structure');
        });

        it('should match HUD elements structure snapshot', () => {
            const hudElements = {
                katamariSize: {
                    element: 'span',
                    id: 'katamari-size',
                    parent: 'game-ui',
                    defaultValue: '2.00m',
                    format: 'decimal with unit'
                },
                katamariSpeed: {
                    element: 'span',
                    id:'katamari-speed',
                    parent: 'game-ui',
                    defaultValue: '0.00m/s',
                    format: 'decimal with unit'
                },
                itemsCollected: {
                    element: 'span',
                    id: 'items-collected',
                    parent: 'game-ui',
                    defaultValue: '0',
                    format: 'integer'
                },
                fps: {
                    element: 'span',
                    id: 'fps',
                    parent: 'game-ui',
                    defaultValue: '--',
                    format: 'integer or placeholder'
                },
                targetSize: {
                    element: 'span',
                    id: 'target-size',
                    parent: 'game-ui',
                    defaultValue: '100.00m',
                    format: 'decimal with unit'
                }
            };

            expect(hudElements).toMatchSnapshot('hud-elements-structure');
        });

        it('should match progress bar structure snapshot', () => {
            const progressContainer = document.getElementById('progress-container');
            const progressBar = document.getElementById('progress-bar');
            
            const progressStructure = {
                container: {
                    tagName: progressContainer.tagName,
                    id: progressContainer.id,
                    children: progressContainer.children.length
                },
                bar: {
                    tagName: progressBar.tagName,
                    id: progressBar.id,
                    parent: progressContainer.id,
                    initialWidth: progressBar.style.width || '0%'
                }
            };

            expect(progressStructure).toMatchSnapshot('progress-bar-structure');
        });

        it('should match control buttons structure snapshot', () => {
            const gyroButton = document.getElementById('gyro-button');
            const debugButton = document.getElementById('debug-button');
            
            const buttonsStructure = {
                gyroButton: {
                    tagName: gyroButton.tagName,
                    id: gyroButton.id,
                    type: gyroButton.type || 'button',
                    defaultText: gyroButton.textContent.trim(),
                    parent: 'game-ui'
                },
                debugButton: {
                    tagName: debugButton.tagName,
                    id: debugButton.id,
                    type: debugButton.type || 'button',
                    defaultText: debugButton.textContent.trim(),
                    parent: 'game-ui'
                }
            };

            expect(buttonsStructure).toMatchSnapshot('control-buttons-structure');
        });

        it('should match power-up status structure snapshot', () => {
            const powerUpStatus = document.getElementById('power-up-status');
            
            const powerUpStructure = {
                tagName: powerUpStatus.tagName,
                id: powerUpStatus.id,
                parent: 'game-ui',
                initialContent: powerUpStatus.innerHTML,
                isEmpty: powerUpStatus.innerHTML === ''
            };

            expect(powerUpStructure).toMatchSnapshot('power-up-status-structure');
        });
    });

    describe('Overlay Structure Snapshots', () => {
        it('should match loading overlay structure snapshot', () => {
            const loadingOverlay = document.getElementById('loading-overlay');
            
            const overlayStructure = {
                tagName: loadingOverlay.tagName,
                id: loadingOverlay.id,
                defaultText: loadingOverlay.textContent.trim(),
                initialDisplay: loadingOverlay.style.display || 'block',
                parent: 'body'
            };

            expect(overlayStructure).toMatchSnapshot('loading-overlay-structure');
        });

        it('should match message overlay structure snapshot', () => {
            const messageOverlay = document.getElementById('message-overlay');
            
            const overlayStructure = {
                tagName: messageOverlay.tagName,
                id: messageOverlay.id,
                initialContent: messageOverlay.innerHTML,
                initialDisplay: messageOverlay.style.display || 'none',
                parent: 'body',
                isEmpty: messageOverlay.innerHTML === ''
            };

            expect(overlayStructure).toMatchSnapshot('message-overlay-structure');
        });

        it('should match controls info structure snapshot', () => {
            const controlsInfo = document.getElementById('controls-info');
            
            const controlsStructure = {
                tagName: controlsInfo.tagName,
                id: controlsInfo.id,
                children: Array.from(controlsInfo.children).map(child => ({
                    tagName: child.tagName,
                    textContent: child.textContent.trim(),
                    hasFormattedText: child.innerHTML.includes('<b>')
                })),
                parent: 'body'
            };

            expect(controlsStructure).toMatchSnapshot('controls-info-structure');
        });
    });

    describe('UI State Variations Snapshots', () => {
        it('should match UI with updated game values snapshot', () => {
            // Update UI elements with game values
            document.getElementById('katamari-size').textContent = '15.75m';
            document.getElementById('katamari-speed').textContent = '8.42m/s';
            document.getElementById('items-collected').textContent = '127';
            document.getElementById('fps').textContent = '60';
            document.getElementById('target-size').textContent = '50.00m';
            document.getElementById('progress-bar').style.width = '31.5%';
            
            const updatedUIState = {
                katamariSize: document.getElementById('katamari-size').textContent,
                katamariSpeed: document.getElementById('katamari-speed').textContent,
                itemsCollected: document.getElementById('items-collected').textContent,
                fps: document.getElementById('fps').textContent,
                targetSize: document.getElementById('target-size').textContent,
                progressWidth: document.getElementById('progress-bar').style.width
            };

            expect(updatedUIState).toMatchSnapshot('ui-with-game-values');
        });

        it('should match UI with active power-ups snapshot', () => {
            const powerUpStatus = document.getElementById('power-up-status');
            powerUpStatus.innerHTML = `
                <div class="power-up">Speed Boost: 4.2s</div>
                <div class="power-up">Magnet: 2.8s</div>
            `;
            
            const powerUpState = {
                hasActivePowerUps: powerUpStatus.innerHTML.trim() !== '',
                powerUpCount: powerUpStatus.querySelectorAll('.power-up').length,
                powerUpContent: powerUpStatus.innerHTML.trim()
            };

            expect(powerUpState).toMatchSnapshot('ui-with-active-power-ups');
        });

        it('should match UI with button states snapshot', () => {
            const gyroButton = document.getElementById('gyro-button');
            const debugButton = document.getElementById('debug-button');
            
            // Set active states
            gyroButton.classList.add('active');
            gyroButton.textContent = 'Gyro ON';
            debugButton.classList.add('active');
            debugButton.textContent = 'Debug ON';
            
            const buttonStates = {
                gyroButton: {
                    isActive: gyroButton.classList.contains('active'),
                    text: gyroButton.textContent,
                    classList: Array.from(gyroButton.classList)
                },
                debugButton: {
                    isActive: debugButton.classList.contains('active'),
                    text: debugButton.textContent,
                    classList: Array.from(debugButton.classList)
                }
            };

            expect(buttonStates).toMatchSnapshot('ui-with-active-buttons');
        });

        it('should match UI with visible overlays snapshot', () => {
            const loadingOverlay = document.getElementById('loading-overlay');
            const messageOverlay = document.getElementById('message-overlay');
            
            // Show overlays
            loadingOverlay.style.display = 'flex';
            loadingOverlay.textContent = 'Generating new world...';
            messageOverlay.style.display = 'block';
            messageOverlay.textContent = 'Level Complete! Well done!';
            
            const overlayStates = {
                loadingOverlay: {
                    isVisible: loadingOverlay.style.display !== 'none',
                    display: loadingOverlay.style.display,
                    content: loadingOverlay.textContent
                },
                messageOverlay: {
                    isVisible: messageOverlay.style.display !== 'none',
                    display: messageOverlay.style.display,
                    content: messageOverlay.textContent
                }
            };

            expect(overlayStates).toMatchSnapshot('ui-with-visible-overlays');
        });
    });

    describe('Responsive UI Structure Snapshots', () => {
        it('should match mobile UI layout considerations snapshot', () => {
            // Simulate mobile viewport
            Object.defineProperty(window, 'innerWidth', { value: 375, writable: true });
            Object.defineProperty(window, 'innerHeight', { value: 667, writable: true });
            
            // Mock mobile media query
            window.matchMedia = vi.fn((query) => ({
                matches: query.includes('max-width: 768px') || query.includes('pointer: coarse'),
                media: query,
                onchange: null,
                addListener: vi.fn(),
                removeListener: vi.fn(),
                addEventListener: vi.fn(),
                removeEventListener: vi.fn(),
                dispatchEvent: vi.fn()
            }));
            
            const mobileUIConsiderations = {
                viewport: {
                    width: window.innerWidth,
                    height: window.innerHeight,
                    isMobile: window.innerWidth <= 768
                },
                touchFriendlyElements: {
                    gyroButton: {
                        id: 'gyro-button',
                        isTouchTarget: true,
                        minTouchSize: '44px' // Recommended minimum
                    },
                    debugButton: {
                        id: 'debug-button',
                        isTouchTarget: true,
                        minTouchSize: '44px'
                    }
                },
                controlsInfo: {
                    includesTouchInstructions: document.getElementById('controls-info').textContent.includes('Swipe'),
                    includesGyroInstructions: document.getElementById('controls-info').textContent.includes('Toggle Gyro')
                }
            };

            expect(mobileUIConsiderations).toMatchSnapshot('mobile-ui-layout-considerations');
        });

        it('should match desktop UI layout considerations snapshot', () => {
            // Simulate desktop viewport
            Object.defineProperty(window, 'innerWidth', { value: 1920, writable: true });
            Object.defineProperty(window, 'innerHeight', { value: 1080, writable: true });
            
            // Mock desktop media query
            window.matchMedia = vi.fn((query) => ({
                matches: query.includes('min-width: 1024px'),
                media: query,
                onchange: null,
                addListener: vi.fn(),
                removeListener: vi.fn(),
                addEventListener: vi.fn(),
                removeEventListener: vi.fn(),
                dispatchEvent: vi.fn()
            }));
            
            const desktopUIConsiderations = {
                viewport: {
                    width: window.innerWidth,
                    height: window.innerHeight,
                    isDesktop: window.innerWidth >= 1024
                },
                keyboardControls: {
                    includesWASDInstructions: document.getElementById('controls-info').textContent.includes('W, A, S, D'),
                    includesArrowKeyInstructions: document.getElementById('controls-info').textContent.includes('Arrow Keys'),
                    includesSpacebarInstructions: document.getElementById('controls-info').textContent.includes('Space')
                },
                debugFeatures: {
                    debugButtonAvailable: !!document.getElementById('debug-button'),
                    fpsDisplayAvailable: !!document.getElementById('fps')
                }
            };

            expect(desktopUIConsiderations).toMatchSnapshot('desktop-ui-layout-considerations');
        });
    });

    describe('UI Accessibility Structure Snapshots', () => {
        it('should match UI accessibility features snapshot', () => {
            const accessibilityFeatures = {
                semanticStructure: {
                    hasMainGameUI: !!document.getElementById('game-ui'),
                    hasControlsInfo: !!document.getElementById('controls-info'),
                    hasProgressIndicator: !!document.getElementById('progress-container')
                },
                interactiveElements: {
                    buttons: Array.from(document.querySelectorAll('button')).map(button => ({
                        id: button.id,
                        text: button.textContent.trim(),
                        hasAccessibleName: button.textContent.trim().length > 0
                    })),
                    clickableOverlays: Array.from(document.querySelectorAll('[id$="-overlay"]')).map(overlay => ({
                        id: overlay.id,
                        isClickable: overlay.id === 'message-overlay'
                    }))
                },
                textContent: {
                    hasInstructionalText: document.getElementById('controls-info').textContent.length > 0,
                    hasFormattedInstructions: document.getElementById('controls-info').innerHTML.includes('<b>'),
                    providesMultipleInputMethods: document.getElementById('controls-info').textContent.includes('Swipe') && 
                                                 document.getElementById('controls-info').textContent.includes('W, A, S, D')
                }
            };

            expect(accessibilityFeatures).toMatchSnapshot('ui-accessibility-features');
        });

        it('should match UI element hierarchy snapshot', () => {
            const uiHierarchy = {
                body: {
                    children: Array.from(document.body.children).map(child => ({
                        tagName: child.tagName,
                        id: child.id,
                        role: child.getAttribute('role') || null,
                        hasChildren: child.children.length > 0,
                        childCount: child.children.length
                    }))
                },
                gameUI: {
                    parent: 'body',
                    children: Array.from(document.getElementById('game-ui').children).map(child => ({
                        tagName: child.tagName,
                        id: child.id || null,
                        hasText: child.textContent.trim().length > 0,
                        isInteractive: child.tagName === 'BUTTON'
                    }))
                }
            };

            expect(uiHierarchy).toMatchSnapshot('ui-element-hierarchy');
        });
    });

    describe('UI Performance Structure Snapshots', () => {
        it('should match UI update performance considerations snapshot', () => {
            const performanceConsiderations = {
                frequentlyUpdatedElements: {
                    katamariSize: {
                        id: 'katamari-size',
                        updateFrequency: 'every frame',
                        isTextContent: true,
                        parentCaching: 'recommended'
                    },
                    katamariSpeed: {
                        id: 'katamari-speed',
                        updateFrequency: 'every frame',
                        isTextContent: true,
                        parentCaching: 'recommended'
                    },
                    fps: {
                        id: 'fps',
                        updateFrequency: 'every second',
                        isTextContent: true,
                        parentCaching: 'recommended'
                    },
                    progressBar: {
                        id: 'progress-bar',
                        updateFrequency: 'on size change',
                        isStyleUpdate: true,
                        property: 'width'
                    }
                },
                staticElements: {
                    controlsInfo: {
                        id: 'controls-info',
                        updateFrequency: 'never',
                        cacheable: true
                    },
                    targetSize: {
                        id: 'target-size',
                        updateFrequency: 'on level change',
                        cacheable: true
                    }
                },
                overlayElements: {
                    loadingOverlay: {
                        id: 'loading-overlay',
                        updateFrequency: 'on state change',
                        displayToggle: true
                    },
                    messageOverlay: {
                        id: 'message-overlay',
                        updateFrequency: 'on game events',
                        displayToggle: true,
                        contentUpdate: true
                    }
                }
            };

            expect(performanceConsiderations).toMatchSnapshot('ui-performance-considerations');
        });
    });
});