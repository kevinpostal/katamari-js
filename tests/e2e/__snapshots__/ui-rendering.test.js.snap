// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UI Rendering Snapshots > Game UI Structure Snapshots > should match HUD elements structure snapshot > hud-elements-structure 1`] = `
{
  "fps": {
    "defaultValue": "--",
    "element": "span",
    "format": "integer or placeholder",
    "id": "fps",
    "parent": "game-ui",
  },
  "itemsCollected": {
    "defaultValue": "0",
    "element": "span",
    "format": "integer",
    "id": "items-collected",
    "parent": "game-ui",
  },
  "katamariSize": {
    "defaultValue": "2.00m",
    "element": "span",
    "format": "decimal with unit",
    "id": "katamari-size",
    "parent": "game-ui",
  },
  "katamariSpeed": {
    "defaultValue": "0.00m/s",
    "element": "span",
    "format": "decimal with unit",
    "id": "katamari-speed",
    "parent": "game-ui",
  },
  "targetSize": {
    "defaultValue": "100.00m",
    "element": "span",
    "format": "decimal with unit",
    "id": "target-size",
    "parent": "game-ui",
  },
}
`;

exports[`UI Rendering Snapshots > Game UI Structure Snapshots > should match complete game UI DOM structure snapshot > complete-game-ui-structure 1`] = `
{
  "children": [
    {
      "children": [
        {
          "className": null,
          "id": "katamari-size",
          "tagName": "SPAN",
          "textContent": "2.00m",
        },
      ],
      "className": null,
      "id": null,
      "tagName": "DIV",
      "textContent": "Size: 2.00m",
    },
    {
      "children": [
        {
          "className": null,
          "id": "katamari-speed",
          "tagName": "SPAN",
          "textContent": "0.00m/s",
        },
      ],
      "className": null,
      "id": null,
      "tagName": "DIV",
      "textContent": "Speed: 0.00m/s",
    },
    {
      "children": [
        {
          "className": null,
          "id": "items-collected",
          "tagName": "SPAN",
          "textContent": "0",
        },
      ],
      "className": null,
      "id": null,
      "tagName": "DIV",
      "textContent": "Items Collected: 0",
    },
    {
      "children": [
        {
          "className": null,
          "id": "fps",
          "tagName": "SPAN",
          "textContent": "--",
        },
      ],
      "className": null,
      "id": null,
      "tagName": "DIV",
      "textContent": "FPS: --",
    },
    {
      "children": [
        {
          "className": null,
          "id": "progress-bar",
          "tagName": "DIV",
          "textContent": "",
        },
      ],
      "className": null,
      "id": "progress-container",
      "tagName": "DIV",
      "textContent": "",
    },
    {
      "children": [
        {
          "className": null,
          "id": "target-size",
          "tagName": "SPAN",
          "textContent": "100.00m",
        },
      ],
      "className": null,
      "id": null,
      "tagName": "DIV",
      "textContent": "Target Size: 100.00m",
    },
    {
      "children": [],
      "className": null,
      "id": "power-up-status",
      "tagName": "DIV",
      "textContent": "",
    },
    {
      "children": [],
      "className": null,
      "id": "gyro-button",
      "tagName": "BUTTON",
      "textContent": "Toggle Gyro",
    },
    {
      "children": [],
      "className": null,
      "id": "debug-button",
      "tagName": "BUTTON",
      "textContent": "Toggle Debug",
    },
  ],
  "id": "game-ui",
  "tagName": "DIV",
}
`;

exports[`UI Rendering Snapshots > Game UI Structure Snapshots > should match control buttons structure snapshot > control-buttons-structure 1`] = `
{
  "debugButton": {
    "defaultText": "Toggle Debug",
    "id": "debug-button",
    "parent": "game-ui",
    "tagName": "BUTTON",
    "type": "submit",
  },
  "gyroButton": {
    "defaultText": "Toggle Gyro",
    "id": "gyro-button",
    "parent": "game-ui",
    "tagName": "BUTTON",
    "type": "submit",
  },
}
`;

exports[`UI Rendering Snapshots > Game UI Structure Snapshots > should match power-up status structure snapshot > power-up-status-structure 1`] = `
{
  "id": "power-up-status",
  "initialContent": "",
  "isEmpty": true,
  "parent": "game-ui",
  "tagName": "DIV",
}
`;

exports[`UI Rendering Snapshots > Game UI Structure Snapshots > should match progress bar structure snapshot > progress-bar-structure 1`] = `
{
  "bar": {
    "id": "progress-bar",
    "initialWidth": "0%",
    "parent": "progress-container",
    "tagName": "DIV",
  },
  "container": {
    "children": 1,
    "id": "progress-container",
    "tagName": "DIV",
  },
}
`;

exports[`UI Rendering Snapshots > Overlay Structure Snapshots > should match controls info structure snapshot > controls-info-structure 1`] = `
{
  "children": [
    {
      "hasFormattedText": true,
      "tagName": "P",
      "textContent": "Use W, A, S, D or Arrow Keys to move the Katamari.",
    },
    {
      "hasFormattedText": true,
      "tagName": "P",
      "textContent": "Swipe on screen for precise mobile control.",
    },
    {
      "hasFormattedText": true,
      "tagName": "P",
      "textContent": "Toggle Gyro for tilt-based movement on supported devices.",
    },
    {
      "hasFormattedText": true,
      "tagName": "P",
      "textContent": "Press R to reset Katamari position. Press Space to generate a new world.",
    },
  ],
  "id": "controls-info",
  "parent": "body",
  "tagName": "DIV",
}
`;

exports[`UI Rendering Snapshots > Overlay Structure Snapshots > should match loading overlay structure snapshot > loading-overlay-structure 1`] = `
{
  "defaultText": "Generating a new universe... âœ¨",
  "id": "loading-overlay",
  "initialDisplay": "block",
  "parent": "body",
  "tagName": "DIV",
}
`;

exports[`UI Rendering Snapshots > Overlay Structure Snapshots > should match message overlay structure snapshot > message-overlay-structure 1`] = `
{
  "id": "message-overlay",
  "initialContent": "",
  "initialDisplay": "none",
  "isEmpty": true,
  "parent": "body",
  "tagName": "DIV",
}
`;

exports[`UI Rendering Snapshots > Responsive UI Structure Snapshots > should match desktop UI layout considerations snapshot > desktop-ui-layout-considerations 1`] = `
{
  "debugFeatures": {
    "debugButtonAvailable": true,
    "fpsDisplayAvailable": true,
  },
  "keyboardControls": {
    "includesArrowKeyInstructions": true,
    "includesSpacebarInstructions": true,
    "includesWASDInstructions": true,
  },
  "viewport": {
    "height": 1080,
    "isDesktop": true,
    "width": 1920,
  },
}
`;

exports[`UI Rendering Snapshots > Responsive UI Structure Snapshots > should match mobile UI layout considerations snapshot > mobile-ui-layout-considerations 1`] = `
{
  "controlsInfo": {
    "includesGyroInstructions": true,
    "includesTouchInstructions": true,
  },
  "touchFriendlyElements": {
    "debugButton": {
      "id": "debug-button",
      "isTouchTarget": true,
      "minTouchSize": "44px",
    },
    "gyroButton": {
      "id": "gyro-button",
      "isTouchTarget": true,
      "minTouchSize": "44px",
    },
  },
  "viewport": {
    "height": 667,
    "isMobile": true,
    "width": 375,
  },
}
`;

exports[`UI Rendering Snapshots > UI Accessibility Structure Snapshots > should match UI accessibility features snapshot > ui-accessibility-features 1`] = `
{
  "interactiveElements": {
    "buttons": [
      {
        "hasAccessibleName": true,
        "id": "gyro-button",
        "text": "Toggle Gyro",
      },
      {
        "hasAccessibleName": true,
        "id": "debug-button",
        "text": "Toggle Debug",
      },
    ],
    "clickableOverlays": [
      {
        "id": "loading-overlay",
        "isClickable": false,
      },
      {
        "id": "message-overlay",
        "isClickable": true,
      },
    ],
  },
  "semanticStructure": {
    "hasControlsInfo": true,
    "hasMainGameUI": true,
    "hasProgressIndicator": true,
  },
  "textContent": {
    "hasFormattedInstructions": true,
    "hasInstructionalText": true,
    "providesMultipleInputMethods": true,
  },
}
`;

exports[`UI Rendering Snapshots > UI Accessibility Structure Snapshots > should match UI element hierarchy snapshot > ui-element-hierarchy 1`] = `
{
  "body": {
    "children": [
      {
        "childCount": 0,
        "hasChildren": false,
        "id": "loading-overlay",
        "role": null,
        "tagName": "DIV",
      },
      {
        "childCount": 9,
        "hasChildren": true,
        "id": "game-ui",
        "role": null,
        "tagName": "DIV",
      },
      {
        "childCount": 4,
        "hasChildren": true,
        "id": "controls-info",
        "role": null,
        "tagName": "DIV",
      },
      {
        "childCount": 0,
        "hasChildren": false,
        "id": "message-overlay",
        "role": null,
        "tagName": "DIV",
      },
    ],
  },
  "gameUI": {
    "children": [
      {
        "hasText": true,
        "id": null,
        "isInteractive": false,
        "tagName": "DIV",
      },
      {
        "hasText": true,
        "id": null,
        "isInteractive": false,
        "tagName": "DIV",
      },
      {
        "hasText": true,
        "id": null,
        "isInteractive": false,
        "tagName": "DIV",
      },
      {
        "hasText": true,
        "id": null,
        "isInteractive": false,
        "tagName": "DIV",
      },
      {
        "hasText": false,
        "id": "progress-container",
        "isInteractive": false,
        "tagName": "DIV",
      },
      {
        "hasText": true,
        "id": null,
        "isInteractive": false,
        "tagName": "DIV",
      },
      {
        "hasText": false,
        "id": "power-up-status",
        "isInteractive": false,
        "tagName": "DIV",
      },
      {
        "hasText": true,
        "id": "gyro-button",
        "isInteractive": true,
        "tagName": "BUTTON",
      },
      {
        "hasText": true,
        "id": "debug-button",
        "isInteractive": true,
        "tagName": "BUTTON",
      },
    ],
    "parent": "body",
  },
}
`;

exports[`UI Rendering Snapshots > UI Performance Structure Snapshots > should match UI update performance considerations snapshot > ui-performance-considerations 1`] = `
{
  "frequentlyUpdatedElements": {
    "fps": {
      "id": "fps",
      "isTextContent": true,
      "parentCaching": "recommended",
      "updateFrequency": "every second",
    },
    "katamariSize": {
      "id": "katamari-size",
      "isTextContent": true,
      "parentCaching": "recommended",
      "updateFrequency": "every frame",
    },
    "katamariSpeed": {
      "id": "katamari-speed",
      "isTextContent": true,
      "parentCaching": "recommended",
      "updateFrequency": "every frame",
    },
    "progressBar": {
      "id": "progress-bar",
      "isStyleUpdate": true,
      "property": "width",
      "updateFrequency": "on size change",
    },
  },
  "overlayElements": {
    "loadingOverlay": {
      "displayToggle": true,
      "id": "loading-overlay",
      "updateFrequency": "on state change",
    },
    "messageOverlay": {
      "contentUpdate": true,
      "displayToggle": true,
      "id": "message-overlay",
      "updateFrequency": "on game events",
    },
  },
  "staticElements": {
    "controlsInfo": {
      "cacheable": true,
      "id": "controls-info",
      "updateFrequency": "never",
    },
    "targetSize": {
      "cacheable": true,
      "id": "target-size",
      "updateFrequency": "on level change",
    },
  },
}
`;

exports[`UI Rendering Snapshots > UI State Variations Snapshots > should match UI with active power-ups snapshot > ui-with-active-power-ups 1`] = `
{
  "hasActivePowerUps": true,
  "powerUpContent": "<div class="power-up">Speed Boost: 4.2s</div>
                <div class="power-up">Magnet: 2.8s</div>",
  "powerUpCount": 2,
}
`;

exports[`UI Rendering Snapshots > UI State Variations Snapshots > should match UI with button states snapshot > ui-with-active-buttons 1`] = `
{
  "debugButton": {
    "classList": [
      "active",
    ],
    "isActive": true,
    "text": "Debug ON",
  },
  "gyroButton": {
    "classList": [
      "active",
    ],
    "isActive": true,
    "text": "Gyro ON",
  },
}
`;

exports[`UI Rendering Snapshots > UI State Variations Snapshots > should match UI with updated game values snapshot > ui-with-game-values 1`] = `
{
  "fps": "60",
  "itemsCollected": "127",
  "katamariSize": "15.75m",
  "katamariSpeed": "8.42m/s",
  "progressWidth": "31.5%",
  "targetSize": "50.00m",
}
`;

exports[`UI Rendering Snapshots > UI State Variations Snapshots > should match UI with visible overlays snapshot > ui-with-visible-overlays 1`] = `
{
  "loadingOverlay": {
    "content": "Generating new world...",
    "display": "flex",
    "isVisible": true,
  },
  "messageOverlay": {
    "content": "Level Complete! Well done!",
    "display": "block",
    "isVisible": true,
  },
}
`;
